### PR TITLE
[RTM] Hotfix for edit/override all in DCG

### PR DESCRIPTION
**Pull request by baumannsven**
This pull request must merged to if DCG has the feature "uniformTime" override/edit all

### CODE REVIEW

**Comment by baumannsven** [comment]
https://github.com/contao-community-alliance/dc-general/pull/359

**Comment by baumannsven** [comment]
@zonky Don´t edit the label of this pull request show in the commit. This isn´t only for uniform time.

**Comment by zonky2** [comment]
na jut

File: src/MetaModels/MetaModel.php [hunk 1]
```diff
@@ -994,11 +995,14 @@ protected function createNewItem($item)
 
     /**
      * {@inheritdoc}
+     *
+     * @SuppressWarnings(PHPMD.Superglobals)
      */
     public function saveItem($objItem)
     {
+        $editInformation = $GLOBALS['container']['dc-general.edit-information'];
```
Comment (dmolineus) [inline src/MetaModels/MetaModel.php]: I don't like it at all. We shouldn't add hidden dependencies. The MM or Item interface should provide a way to handle it.

I suggest following. As we are still in alpha, we could break the API without bumping the major version.

```php

public function saveItem($objItem, $setTimestamp = true);

```

Then the DCG data provider could handle it. 

After merging this PR it wouldn't be possible to use MM outside of the DCG context!

I'm also not sure if the edit information class is needed at all. The DCG models provide a way for meta data. But let's discuss it in the DCG related PR.

**Comment by baumannsven** [comment]
@dmolineus @discordier I have remove the hidden dependency

**Comment by dmolineus** [comment]
This change request modifies interfaces so we have backward compatibility issues here. Probably better to add it only for MM 2.1. Wdyt @discordier?

**Comment by discordier** [comment]
As there are default values on the interface, any existing concrete class should still work.

However should trigger an deprecation notice if not given.

File: src/MetaModels/DcGeneral/Data/Driver.php [hunk 2]
```diff
@@ -569,13 +569,15 @@ public function resetFallback($strField)
...
+     * @param ModelInterface $objItem   The model to save back.
+     *
+     * @param int            $timestamp Optional the timestamp.
      *
      * @return ModelInterface The passed model.
      *
      * @throws \RuntimeException When an incompatible item was passed, an Exception is being thrown.
      */
-    public function save(ModelInterface $objItem)
+    public function save(ModelInterface $objItem, $timestamp = 0)
```
Comment (discordier) [inline src/MetaModels/DcGeneral/Data/Driver.php]: The tstamp should be passed to `$objItem->getItem()->save();` below!

Comment (discordier) [reply src/MetaModels/DcGeneral/Data/Driver.php]: We also should trigger a
```php
@trigger_error('Not passing a timestamp has been deprecated and will cause an error in MetaModels 3', E_USER_DEPRECATED);
```

File: src/MetaModels/DcGeneral/Data/Driver.php [hunk 3]
```diff
@@ -593,13 +595,15 @@ public function save(ModelInterface $objItem)
...
+     * @param CollectionInterface $objItems  The collection containing all items to be saved.
+     *
+     * @param int                 $timestamp Optional the timestamp.
      *
      * @return void
      *
      * @throws \RuntimeException When an incompatible item was passed.
      */
-    public function saveEach(CollectionInterface $objItems)
+    public function saveEach(CollectionInterface $objItems, $timestamp = 0)
```
Comment (discordier) [inline src/MetaModels/DcGeneral/Data/Driver.php]: We also should trigger a
```php
@trigger_error('Not passing a timestamp has been deprecated and will cause an error in MetaModels 3', E_USER_DEPRECATED);
```

File: src/MetaModels/DcGeneral/Data/Driver.php [hunk 4]
```diff
@@ -593,13 +595,15 @@ public function save(ModelInterface $objItem)
...
     {
         foreach ($objItems as $objItem) {
             $this->save($objItem);
```
Comment (discordier) [inline src/MetaModels/DcGeneral/Data/Driver.php]: The timestamp should get passed here.

File: src/MetaModels/Item.php [hunk 5]
```diff
@@ -339,12 +339,14 @@ public function getSiblings($objFilter)
...
      * Save the current data for every attribute to the data sink.
      *
+     * @param int $timestamp Optional the timestamp.
+     *
      * @return void
      */
-    public function save()
+    public function save($timestamp = 0)
     {
         $objMetaModel = $this->getMetaModel();
```
Comment (discordier) [inline src/MetaModels/Item.php]: We also should trigger a
```php
@trigger_error('Not passing a timestamp has been deprecated and will cause an error in MetaModels 3', E_USER_DEPRECATED);
```

File: src/MetaModels/MetaModel.php [hunk 6]
```diff
@@ -1006,11 +1006,13 @@ protected function createNewItem($item)
 
     /**
      * {@inheritdoc}
+     *
+     * @SuppressWarnings(PHPMD.Superglobals)
      */
-    public function saveItem($objItem)
+    public function saveItem($objItem, $timestamp = 0)
     {
```
Comment (discordier) [inline src/MetaModels/MetaModel.php]: We also should trigger a
```php
@trigger_error('Not passing a timestamp has been deprecated and will cause an error in MetaModels 3', E_USER_DEPRECATED);
```

File: src/MetaModels/DcGeneral/Data/Driver.php (hunk 2 above)
Comment (baumannsven) [reply src/MetaModels/DcGeneral/Data/Driver.php]: I have taken note of this

File: src/MetaModels/DcGeneral/Data/Driver.php (hunk 3 above)
Comment (baumannsven) [reply src/MetaModels/DcGeneral/Data/Driver.php]: I have taken note of this

File: src/MetaModels/DcGeneral/Data/Driver.php (hunk 4 above)
Comment (baumannsven) [reply src/MetaModels/DcGeneral/Data/Driver.php]: I have taken note of this

File: src/MetaModels/Item.php (hunk 5 above)
Comment (baumannsven) [reply src/MetaModels/Item.php]: I have taken note of this

File: src/MetaModels/MetaModel.php (hunk 6 above)
Comment (baumannsven) [reply src/MetaModels/MetaModel.php]: I have taken note of this

**Comment by baumannsven** [comment]
@discordier I put the deprecated notification in an if statement and change the default timestamp value from 0 to null

**Comment by discordier** [comment]
@dmolineus, @stefanheimes could you re-evaluate this PR and either approve or request changes?

**Pull request merged**